File: ios/Classes/SystemInfoAgentPlugin.swift
//
// Created by Thurain Hein on 22/08/2025.
//

import Foundation
import Flutter
import UIKit
import WebKit

@objc(SystemInfoAgentPlugin)
public class SystemInfoAgentPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "system_info_agent", binaryMessenger: registrar.messenger())
        let instance = SystemInfoAgentPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getProperties":
            let userAgent = getDalvikUserAgent()
            getWebViewUserAgent { webViewUserAgent in
                       let properties: [String: String] = [
                           "userAgent": userAgent,
                           "webViewUserAgent": webViewUserAgent
                       ]
                       result(properties)
                   }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    

    // Dalvik style userAgent (fake format for iOS)
    private func getDalvikUserAgent() -> String {
        let osVersion = UIDevice.current.systemVersion
        let model = UIDevice.current.model
        let id = ProcessInfo.processInfo.operatingSystemVersionString
        return "Dalvik/2.1.0 (Linux; U; iOS \(osVersion); \(model) Build/\(id))"
    }

    // WebView user agent (sync fetch using DispatchSemaphore)
    private func getWebViewUserAgent(completion: @escaping (String) -> Void) {
        let webView = WKWebView(frame: .zero)

        if let ua = webView.value(forKey: "userAgent") as? String {
            print("WebView UA via value(forKey:) ------ \(ua)")
            completion(ua)
        } else if let ua = webView.customUserAgent {
            print("WebView UA via customUserAgent ----- \(ua)")
            completion(ua)
        } else {
            webView.evaluateJavaScript("navigator.userAgent") { (result, error) in
                if let ua = result as? String {
                    print("WebView UA via evaluateJavaScript: \(ua)")
                    completion(ua)
                } else {
                    print("WebView UA fetch failed, error: \(String(describing: error))")
                    completion("Unknown")
                }
            }
        }
    }
}

File: android/src/main/kotlin/com/thurainhein/system_info_agent/SystemInfoAgentPlugin.kt
package com.thurainhein.system_info_agent

import android.os.Build
import android.webkit.WebSettings
import android.webkit.WebView
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class SystemInfoAgentPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: android.content.Context

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "system_info_agent")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        when (call.method) {
            "getProperties" -> {
                val map = mapOf(
                    "userAgent" to getDalvikUserAgent(),
                    "webViewUserAgent" to getWebViewUserAgent()
                )
                result.success(map)
            }
            else -> result.notImplemented()
        }
    }

    // Dalvik format userAgent
    private fun getDalvikUserAgent(): String {
        val version = System.getProperty("java.vm.version") ?: "2.1.0"
        val os = "Linux; U; Android ${Build.VERSION.RELEASE}; ${Build.MODEL} Build/${Build.ID}"
        return "Dalvik/$version ($os)"
    }

    // Actual WebView user agent
    private fun getWebViewUserAgent(): String {
        return WebSettings.getDefaultUserAgent(context)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}





